executor/builder.go

```diff
@@ -1236,15 +1236,79 @@ func (b *executorBuilder) buildUpdate(v *plan.Update) Executor {
 		b.err = errors.Trace(b.err)
 		return nil
 	}
+	columns2Handle := buildColumns2Handle(v.Schema(), tblID2table)
 	updateExec := &UpdateExec{
-		baseExecutor: newBaseExecutor(b.ctx, nil, v.ExplainID(), selExec),
-		SelectExec:   selExec,
-		OrderedList:  v.OrderedList,
-		tblID2table:  tblID2table,
+		baseExecutor:   newBaseExecutor(b.ctx, nil, v.ExplainID(), selExec),
+		SelectExec:     selExec,
+		OrderedList:    v.OrderedList,
+		tblID2table:    tblID2table,
+		columns2Handle: columns2Handle,
 	}
 	return updateExec
 }
 
+// Columns2HandleEntry represents an mapper from column index to handle index.
+type Columns2HandleEntry struct {
+	start, end int
+	handleIdx  int
+}
+
+// Columns2Handle attaches the methods of sort.Interface to []Columns2HandleEntry sorting in increasing order.
+type Columns2Handle []Columns2HandleEntry
+
+// Len implements sort.Interface#Len.
+func (c Columns2Handle) Len() int {
+	return len(c)
+}
+
+// Len implements sort.Interface#Swap.
+func (c Columns2Handle) Swap(i, j int) {
+	c[i], c[j] = c[j], c[i]
+}
+
+// Len implements sort.Interface#Less.
+// let ranges first sorted by `start` increasing order, and then sorted by `end` increasing order if `start` are equal.
+func (c Columns2Handle) Less(i, j int) bool {
+	if c[i].start == c[j].start {
```

Comment:
This situation should never happen?

Reply:
I'm not very sure about this...so add this protection.... 23333